pennylane/gradients/finite_difference.py

```diff
@@ -188,15 +197,17 @@ def _expand_transform_finite_diff(
     validate_params=True,
 ) -> (Sequence[qml.tape.QuantumTape], Callable):
     """Expand function to be applied before finite difference."""
-    expanded_tape = expand_invalid_trainable(tape)
-
-    def null_postprocessing(results):
-        """A postprocesing function returned by a transform that only converts the batch of results
-        into a result for a single ``QuantumTape``.
-        """
-        return results[0]
-
-    return [expanded_tape], null_postprocessing
+    [new_tape], postprocessing = qml.devices.preprocess.decompose(
+        tape,
+        stopping_condition=_finite_diff_stopping_condition,
+        skip_initial_state_prep=False,
+        name="param_shift",
+    )
+    if new_tape is tape:
+        return [tape], postprocessing

```

Comment:
If it's the old tape, are we guaranteed to always have valid `trainable_params` at this point?

Reply:
I would say we are rarely, if ever, guaranteed to have valid `trainable_params`.  But at least this way we keep the trainable params that we had before, what ever those were.

Reply:
I see. I was thinking about a case where changing trainability will change whether or not `decompose` changes the tape. Then this tape expansion will have valid trainable parameters in one case but not in the other :grimacing: But I agree that there is no system to adhere to, so fine to leave it as is.